### PR TITLE
ci: refactor external dep build/install scripts.

### PR DESCRIPTION
**Pull request by htuch**
The motivation for this is that we would like to allow developers to perform local installation of
Envoy dependencies as prebuilts, similar to CI.

build_and_install_deps.sh now does the build/install into $THIRDPARTY_{SRC,BUILD} via recursive
make. This provides a significant speed boost (2 minutes total build time for all dependencies on my
workstation) and the top-level make is able to keep track of when stuff needs to be rebuilt as
dependencies change over time (see comment at the top of Makefile).

Verified with {debug, coverage, bazel.debug} runs locally against the resulting CI Docker image.

### CODE REVIEW

File: ci/build_container/Makefile [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# The individual build recipe scripts must contain sufficient information (e.g.  SHA, URL, repo,
+# version number, etc.) to uniquely identify the revision of the upstream dependency.  This allows
+# make to pick up changes with a simple direct dependency on the build recipe.
+# TODO(htuch): These dependencies should be installed into distinct --prefix directories, so we can
+# easily blow away the lib/ and include/ install directories when we upgrade version.
+
+all: $(THIRDPARTY_DEPS)/libevent.dep \
```
Comment (mattklein123) [inline ci/build_container/Makefile]: nit: Can we fix the whitespace indent in this file. It's making my OCD flair up.

File: ci/build_container/build_recipes/boringssl.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+git clone https://boringssl.googlesource.com/boringssl
+cd boringssl
+git reset --hard b87c80300647c2c0311c1489a104470e099f1531
+cmake .
+make
```
Comment (mattklein123) [inline ci/build_container/build_recipes/boringssl.sh]: For full job server support aren't you supposed to do ```$(MAKE)``` ? (Similar elsewhere)

File: ci/build_container/Makefile (hunk 1 above)
Comment (htuch) [inline ci/build_container/Makefile]: Done

File: ci/build_container/build_recipes/boringssl.sh (hunk 2 above)
Comment (htuch) [inline ci/build_container/build_recipes/boringssl.sh]: By prefixing the shell invocation line with +, the top-level make puts the make coordination flags in the environment, so no need to do anything special in the script.

**Pull request merged**